hazelcast/src/main/java/com/hazelcast/sql/impl/expression/CaseExpression.java

```diff
@@ -0,0 +1,156 @@
+    private Expression<?>[] thenExpressions;
+    private Expression<?> elseExpression;
+    private QueryDataType resultType;
+
+    public CaseExpression() {
+    }
+
+    private CaseExpression(Expression<Boolean>[] whenExpressions,
+                           Expression<?>[] thenExpressions,
+                           Expression<?> elseExpression,
+                           QueryDataType resultType) {
+        this.whenExpressions = whenExpressions;
+        this.thenExpressions = thenExpressions;
+        this.elseExpression = elseExpression;
+        this.resultType = resultType;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T> CaseExpression<T> create(Expression<?>[] operands, QueryDataType resultType) {
+        int branchesSize = operands.length / 2;
+        Expression<Boolean>[] whenExpressions = new Expression[branchesSize];
+        Expression<?>[] thenExpressions = new Expression[branchesSize];
+        int index = 0;
+        for (int i = 0; i < operands.length - 1; i += 2) {
+            whenExpressions[index] = (Expression<Boolean>) operands[i];
+            thenExpressions[index] = operands[i + 1];
+            index += 1;
+        }
+        return new CaseExpression<>(whenExpressions, thenExpressions, operands[operands.length - 1], resultType);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_CASE;
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeInt(whenExpressions.length);
+
+        for (int i = 0; i < whenExpressions.length; i++) {
+            out.writeObject(whenExpressions[i]);
+            out.writeObject(thenExpressions[i]);
+        }
+
+        out.writeObject(elseExpression);
+        out.writeObject(resultType);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        int branchesSize = in.readInt();
+
+        whenExpressions = new Expression[branchesSize];
+        thenExpressions = new Expression[branchesSize];
+
+        for (int i = 0; i < branchesSize; i++) {
+            whenExpressions[i] = in.readObject();
+            thenExpressions[i] = in.readObject();
+        }
+
+        elseExpression = in.readObject();
+        resultType = in.readObject();
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T eval(Row row, ExpressionEvalContext context) {
+        for (int i = 0; i < whenExpressions.length; i++) {
+            Expression<Boolean> condition = whenExpressions[i];
+
+            Boolean conditionHolds = condition.eval(row, context);
+            if (TernaryLogic.isTrue(conditionHolds)) {
+                return (T) thenExpressions[i].eval(row, context);
+            }
+        }
+
+        return (T) elseExpression.eval(row, context);
```

Comment:
Please note that the `else` part is optional in the `CASE` operator. We need null-checks here, as well as associated tests.

Reply:
`doesntMatchWhen_andNoElseBranch` integration test https://github.com/hazelcast/hazelcast/pull/18067/files#diff-d5f2f622f953b687afaf60419e612a9b48b2e5c439a09e5b382a7b4064101804R82 should cover situation when `else` is omitted and `when` evaluates to `false`.